src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterEnum.cs

```diff
@@ -46,11 +55,106 @@ public override bool TryRead(Type valueType, ref Utf8JsonReader reader, out TVal
                     value = (TValue)Enum.ToObject(valueType, ulongValue);
                     return true;
                 }
+                else if (reader.TryGetInt64(out long fallback) && fallback == -1)
+                {
+                    value = (TValue)Enum.ToObject(valueType, fallback);
+                    return true;
+                }
+
+                value = default;
+                return false;
+            }
+
+            if (s_isInt64)
+            {
+                if (reader.TryGetInt64(out long ulongValue))
+                {
+                    value = (TValue)Enum.ToObject(valueType, ulongValue);
+                    return true;
+                }
+
+                value = default;
+                return false;
+            }
+
+            if (s_isUint32)
+            {
+                if (reader.TryGetUInt32(out uint uintValue))
+                {
+                    value = (TValue)Enum.ToObject(valueType, uintValue);
+                    return true;
+                } 
+                else if (reader.TryGetInt64(out long fallback) && fallback == -1)
+                {
+                    value = (TValue)Enum.ToObject(valueType, fallback);
+                    return true;
+                }
+
+                value = default;
+                return false;
+            }
+
+            if (s_isInt32)
+            {
+                if (reader.TryGetInt32(out int intValue))
+                {
+                    value = (TValue)Enum.ToObject(valueType, intValue);
+                    return true;
+                } 
+
+                value = default;
+                return false;
+            }
+
+            if (s_isUshort)
+            {
+                if (reader.TryGetUInt32(out uint uintValue) && uintValue >= ushort.MinValue && uintValue <= ushort.MaxValue)
+                {
+                    value = (TValue)Enum.ToObject(valueType, uintValue);
+                    return true;
+                }
+                else if (reader.TryGetInt64(out long fallback) && fallback == -1)
+                {
+                    value = (TValue)Enum.ToObject(valueType, fallback);
+                    return true;
+                }
+
+                value = default;
+                return false;
+            }
+
+            if (s_isshort)
+            {
+                if (reader.TryGetInt32(out int intValue) && intValue >= short.MinValue && intValue <= short.MaxValue)
```

Comment:
Feel free to link to this issue on the reader for now (as a comment), which will remove the need for custom checks. https://github.com/dotnet/corefx/issues/36125